README.md
# Siccuracy #

Stefan's imputation accuracy package. And with helper functions for working with AlphaSuite's SNP files.
Implemented in Fortran for speed.

See also script [runAlphaImpute](runAlphaImpute/)

## What is this repository for? ###

This package was developed for working with SNP files in the format used in [AlphaSuite](https://bitbucket.org/tutorials/markdowndemo). The format of these consists of first column with sample ID (Individual or Animal ID), and subsequent column counting minor alleles at each loci: 

```
1001      0    0    1    2    1    2
1002   0.05 0.50 1.50 1.80 0.95    9
```

In this example, the file consists of two individuals (1001 and 1002); first individual is has genotypes (`0`,`0`,`1`,`2`,`1`,`2`) for six loci. The second individual has been imputed, and the genotypes are given as real values of best guess. The last genotype was not sufficiently imputed and therefore set as missing with value ``9``.

Requirements for format: Space separated; number of digits is unimportant.

## To use ##

In R, load package with `library(Siccuracy)`. There exists a number of functions:

* Calculate imputation accuracy (i.e. correlation between two genotype matrices): `imputation_accuracy3`
* Calculate hetereozygosity: `heterozygosity`
* Helper functions: `get_ncol`, `get_firstcolumn`, `get_nlines`
* Combining files: `rowconcatenate` and `mergeChips`
* Reading and writing files: `write.snps`, `read.snps`
* Convert phased files to genotype files: `phasotogeno`

**Most functions operate directly with files, not matrices loaded into memory!**


## How do I get set up? ###


Download source folder, from command line run

```
#!
R CMD INSTALL Siccuracy
```

or, (untested)


```
#!r

library(devtools)
install_github(repo='stefanedwards/Siccuracy', ref = "master")
```

## I found a bug ##

Lucky you.

Please report it using the issues tool.

## Can' some of these function be done in bash?

Presumable.

The `rowconcatenate` can be done by using `paste` in bash -- except it requires that you first strip off
first column of all files except the first.


NEWS
# 0.8.0

  * Function renaming:
    * `cbind_SNPs` --> `cbind_snp_files`
    * `rbind_SNPs` --> `rbind_snp_files`
    * `mask_SNPs` --> `mask_snp_file`

  * Added support for reading SHAPEIT haps/sample files (`read.haps`)
  * Added support for reading Oxford formatted files (`read.oxford`).
  
  * `imputation_accuracy`: Standardizing with `p` is now scaling with the
    *squareroot* of the variance of allele frequency. Before this, it was the
    variance.

# 0.7.1

  * Fortran routines updated to be able to handle integer IDs exceeding 
    2147483647. Passing larger integers back to R is however still not possible.
  * Added support for vcf files with package 'vcfR', but require the genotypes
    to be loaded with `read.vcfR`.
    Support is on `write.snps`, and `imputation_accuracy`.
  * `imputation_accuracy` has slight modified behaviour:
    It now dispatches to either files or matrices depending on first argument.
    Correct.pct is calculated as # correct / (elements - true missing).
    Tolerance is limited to 4 digits. I.e. a tolerance below 1e-4 is effectively
    ignored.

# 0.5.0

  * Preparing package for publication.
  * get_firstcolumn now works with several data types, but default use is still integer.

# 0.3.2

  * Added function `mergeChips` for merging multiple SNP chips into a single file.
  * Added function `read.snps` for easy reading SNP files.

# 0.3.1

  * Added function `rowconcatenate`, albeit without test routines.
  * Updated phasotogeno to handle numeric values instead of just integers.
  * phasotogeno_int is the integer version of phasotogeno (where the latter does not work with heterozygosity).
  * Bug fix: Now standardises by standard deviation, not variance.

# 0.3.0

  * Added NEWS file.
  * In `imputation_accuracy3`, row and columns correlations that are `Inf`, `-Inf`, or `NaN` are set to `NA`.

# 0.2.0

  * Added Heterozygosity function.


R/Siccuracy.R
#' Stefan's imputation accuracy package
#' 
#' Calculation of imputation accuracies and similar allele stats.
#' 
#' This package was developed for working with SNP files in the format used in \href{http://www.alphagenes.roslin.ed.ac.uk/}{AlphaSuite}. 
#' The format of these consists of first column with \emph{integer} sample ID (Individual or Animal ID),
#' and subsequent column counting minor alleles at each loci:
#' 
#' \tabular{lcccccc}{
#'  1001  \tab    0   \tab   0    \tab 1    \tab  2   \tab  1   \tab   2 \cr
#'  1002    \tab 0.05   \tab 0.50   \tab 1.50   \tab 1.80   \tab 0.95     \tab 9 \cr
#' }
#' 
#' In this example, the file consists of two individuals (\code{1001} and \code{1002}); 
#' first individual has genotypes (\code{0}, \code{0}, \code{1}, \code{2}, \code{1}, \code{2}) for six loci. 
#' The second individual has been imputed, and the genotypes are given as real values of best guess. 
#' The last genotype was not sufficiently imputed and therefore set as missing with value \code{9}.
#'
#' Requirements for format: Space or tab separated; number of digits is unimportant.
#' 
#' @section Overview:
#' 
#' \itemize{
#'   \item \strong{Converters} \code{\link{convert_phases}}, \code{\link{convert_plink}}, \code{\link{convert_plinkA}}.
#'   \item \strong{File info} \code{\link{get_nlines}}, \code{\link{get_firstcolumn}}, \code{\link{get_ncols}}.
#'   \item \strong{Imputation accuracies} \code{\link{imputation_accuracy}} correlations between matrices, and tabulating correctly imputed genotypes.
#'   \item \strong{Heterozygosity} \code{\link{heterozygosity}} counting alleles.
#'   \item \strong{Mutators} \code{\link{cbind_snp_files}}, \code{\link{rbind_snp_files}} for combining multiple files, \code{\link{mask_SNPs}} for masking SNPs for simulating different density SNP chips.
#'   \item \strong{Reading and writing} \code{\link{write.snps}}, \code{\link{read.snps}},
#'                                      \code{\link{read.haps}}, \code{\link{read.oxford}},
#'                                      \code{\link[=VCF format]{read.vcf}}.
#' }
#' 
#' @section IDs:
#' 
#' The AlphaImpute format \strong{must} have integer IDs. 
#' In this software package, we have choosen an arbitrary upper limit of 20 digits for IDs.
#' Exceed this limit and the ID will be lumped together with the following genotype.
#' 
#' Regarding \emph{repeated} IDs, it is best to avoid these.
#' 
#' @section Output format:
#' 
#' Functions that write new files may have a pair of arguments (\code{int} and \code{format}) 
#' that specifies whether the outputted format are integers (\code{int=TRUE}) or have decimals.
#' For more information on how to specify the format, see \link{parseformat}.
#' 
#' @section Other formats:
#' 
#' \strong{Oxford} format is covered with \code{\link{read.oxford}}
#' and a method dispatch for imputation accuracy exists for this too.
#' 
#' \strong{SHAPEIT}'s haps/sample format is covered with \code{\link{read.haps}} 
#' and a method dispatch for imputation accuracy exists for this too.
#' 
#' Currently, \strong{VCF} objects by package \code{vcfR} are supported for functions
#' \code{\link[=write.snps.vcfR]{write.snps}} and \code{\link[=imputation_accuracy.vcfR]{imputation_accuracy}}.
#' For more information, see \link{VCF format}.
#' 
#' 
#' @docType package
#' @name Siccuracy
#' @aliases AlphaImpute-format
#' @useDynLib Siccuracy
#' @author Stefan McKinnon Edwards <sme@@iysik.com>
NULL
